mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -0,0 +1,310 @@
+    RankedTensorType resultType = op.getResultTypes().front().cast<RankedTensorType>();
+
+    // Loop over the scalars to create a folded circuit per factor
+    Value c0 = rewriter.create<index::ConstantOp>(loc, 0);
+    Value c1 = rewriter.create<index::ConstantOp>(loc, 1);
+    Value size = rewriter.create<index::ConstantOp>(loc, sizeInt);
+    // Initialize the results as empty tensor
+    Value results =
+        rewriter.create<tensor::EmptyOp>(loc, resultType.getShape(), resultType.getElementType());
+    Value resultValues =
+        rewriter
+            .create<scf::ForOp>(
+                loc, c0, size, c1, /*iterArgsInit=*/results,
+                [&](OpBuilder &builder, Location loc, Value i, ValueRange iterArgs) {
+                    std::vector<Value> newArgs{op.getArgs().begin(), op.getArgs().end()};
+                    SmallVector<Value> index = {i};
+                    Value scalarFactor =
+                        builder.create<tensor::ExtractOp>(loc, scalarFactors, index);
+                    newArgs.push_back(scalarFactor);
+                    Value resultValue =
+                        builder.create<func::CallOp>(loc, foldedCircuit, newArgs).getResult(0);
+
+                    Value resultInserted =
+                        builder.create<tensor::InsertOp>(loc, resultValue, iterArgs.front(), index);
+                    builder.create<scf::YieldOp>(loc, resultInserted);
+                })
+            .getResult(0);
+
+    // Replace the original results
+    rewriter.replaceOp(op, resultValues);
+}
+
+FlatSymbolRefAttr ZneLowering::getOrInsertFoldedCircuit(Location loc, PatternRewriter &rewriter,
+                                                        mitigation::ZneOp op, Type scalarType)
+{
+    MLIRContext *ctx = rewriter.getContext();
+
+    OpBuilder::InsertionGuard guard(rewriter);
+    ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+    std::string fnFoldedName = op.getCallee().str() + ".folded";
+
+    if (moduleOp.lookupSymbol<func::FuncOp>(fnFoldedName)) {
+        return SymbolRefAttr::get(ctx, fnFoldedName);
+    }
+
+    // Original function
+    func::FuncOp fnOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+    TypeRange originalTypes = op.getArgs().getTypes();
+    Type qregType = quantum::QuregType::get(rewriter.getContext());
+
+    // Set insertion in the module
+    rewriter.setInsertionPointToStart(moduleOp.getBody());
+    // Quantum Alloc function
+    FlatSymbolRefAttr quantumAllocRefAttr = getOrInsertQuantumAlloc(loc, rewriter, op);
+    func::FuncOp fnAllocOp =
+        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, quantumAllocRefAttr);
+
+    // Get the number of qubits
+    quantum::AllocOp allocOp = *fnOp.getOps<quantum::AllocOp>().begin();
+    std::optional<int64_t> numberQubitsOptional = allocOp.getNqubitsAttr();
+    int64_t numberQubits = numberQubitsOptional.value_or(0);
```

Comment:
Is there any implicit casting involved?

Reply:
I think the zero is casted indeed if there is no value.